v4.0.0-rc.7

pushrbx opened the pull request:
## Changelog
- fully working search engine (no more buggy results)
- updated api docs annotations
- added full request validation
- added data transfer objects for requests
- modernized jikan config (e.g. cache ttl): the config is now passed around as an object instead of an array
- added typed enums for each enum specified in the api docs
- added unit tests
- fixed filtering in seasons API endpoints
- fixed bugs around top endpoints
- fixed top reviews endpoint -- please check the new api docs as the parameters have changed
- housekeeping - removed old things
- fixed status code inconsistency with upstream MAL server - e.g. if a request is sent to MAL in case of a cache miss, and the response failed from MAL, then the page is not cached and the response code either returned from MAL to the client or transformed to a different one.
- request validation errors will now produce a json response with clear error messages
- reduced typesense's memory usage: changed the indexing of the data by changing the type of the `mal_id` attribute to `int` instead of `string`, so we can sort by `mal_id` when looking for items and ordering by `mal_id` when it's required.
- all search endpoints will now properly override search engine options for ordering results in the response
- improved roadrunner integration
- disabled micro caching for automated tests
- fixed an issue with micro caching where the data is asked for multiple times
- upgraded dependencies
- improved the container images: bumped versions, and added xdebug if you ever want to debug the container
- fixed #291
- fixed #292
- fixed #297 
- fixed #298 
- fixed #299 
- fixed #309 
- fixed #323 
- fixed #327 
- fixed #332 
- fixed #334 
- fixed #335 
- added CORS support in case of self-hosting: out of the box the app sets the correct headers.

## Task list

- [x] Add dtos
- [x] Refactor controllers
- [x] Add unit tests
- [x] Refactor existing integration tests to use dummy data
- [x] Refactor existing integration tests to avoid calls to MAL during tests
- [x] Add documentation about changes, how the class structure have changed
- [x] Add documentation about laravel app container lifecycles when using PSR-7 workers and roadrunner application server

## New dependencies

- https://spatie.be/docs/laravel-data/v2/introduction -- used for request validation
- https://github.com/spatie/laravel-enum -- used for request validation
- [haydenpierce/class-finder](https://gitlab.com/hpierce1102/ClassFinder) -- used for model lookups during test runner bootstrap

## Architecture changes

This PR changes the core architecture of the app. Instead of having some "query builder" classes which are used throughout the methods of "controller" classes, there is now a mediator pattern in place, where the controller uses a mediator instance to handle the request. This resulted in pairs of a request handler and a strongly typed request object aka data transfer object.

![image](https://user-images.githubusercontent.com/6832715/218834276-fbdc8be9-39af-426f-9b52-e196876d9666.png)

![image](https://user-images.githubusercontent.com/6832715/218836127-2f9b9cbb-0c51-4955-ab4d-2f2645a1c4d3.png)


### Changes over time (since June, 2022)

![image](https://user-images.githubusercontent.com/6832715/218832621-3a105ef8-bac6-4abb-b3da-db3228da276d.png)

### Request handlers

These classes are configured in the `AppServiceProvider` class, which tells the Laravel container how to build these classes.
These are then fed to the singleton mediator instance, which looks them up by their `requestClass` method.

> Trivia: I have a strong C#/ASP.NET background, there you can use generic classes, but in PHP you can only annotate the classes as generic class, and have their type parameter as return value in one of their methods.

### Data transfer objects

Thanks to [Spatie's laravel-data package](https://spatie.be/docs/laravel-data/v2/introduction), all request validation is done by creating specialisations of their `Data` class. This also meant that the php version should be greater or equals to 8.1, because the package encourages the usage of [attributes/decorators](https://www.php.net/manual/en/language.attributes.overview.php), which was introduced in php 8.1.

### Unit of Work and Repositories "pattern"

There is already some sort of "unit of work" pattern in place in Laravel. This new one introduced by the PR, is something similar:
Have a class representation of all possible tables, which are queryable. This will help to have an overview of all the data the app is dealing with.

Repositories are just abstractions over the model classes and they are encapsulating the special queries for those models. There is also a special repository interface implementation which is meant to represent a queryable collection from mongodb. It takes a collection name string as constructor argument.

These two additions will encapsulate the data access in the app. Worth noting here that by data access we mean the "cached" data from upstream.

### Consolidated way to do cache lookups

Jikan essentially is a big cache, where some of the models are indexed and searchable. If the record in the database is stale, then a fresh version is scraped from upstream. There is now a "service" in the code which does the cache validation and the cache update: `CachedScraperService`.
This service uses the repositories to lookup data, and if the data is stale, it uses the `MalClient` class instance from the Jikan lib to fetch a new version from upstream.

### More dry code

There are now less repetitions. The core logic of the app are now centralised across a few core classes:
- `DefaultCachedScraperService` class for scraping from upstream if content in DB is stale, otherwise it returns cached content from DB.
- `RequestHandlerWithScraperCache` base class for all those request handlers which are returning not indexed data.
- `ItemLookupHandler` base class for all those request handlers which are returning indexed data by a key. (Id, username etc.)
- `SearchRequestHandler` base class for all those request handlers which allows the clients/users to search indexed data.

## RoadRunner quirks

There are some consequences of using roadrunner as application server: The lifecycles of objects are a bit different.
Under RoadRunner the application lives as a PSR-7 worker, there can be many running workers in parallel, where each of them is a dedicated process. These processes are managed by RoadRunner. The application will be loaded and boostrapped only one  time when the Worker process starts up. This means that the whole application can be stored and kept in memory, instead of the usual per request bootstrap and loading into memory procedure. (E.g. apache web server + modfpm / modphp or nginx + fpm) This also means that the instances in the IOC container remain in memory, and they are only unloaded when roadrunner stops, or it reduces the number of workers.
The IOC container is the core concept of Lumen/Laravel's structure. It can be used for application binding, singleton, resolving applications, dependency injections and more. Without RoadRunner and Workers you don't have to care about the usages of `global properties`, `static class`, `singletons`. With RoadRunner now it matters.

For example, auth in Laravel is registered as singleton. Here is the partial code of SessionGuard:
```php
if (! is_null($this->user)) {
    return $this->user;
}
```
It looks pretty normal and reasonable for caching resolved user object. But in our case, this is a fatal problem. That will cause you getting wrong user object after your first user resolving.

All of this is resolved through the [pushrbx/lumen-roadrunner](https://github.com/pushrbx/lumen-roadrunner/) package, which provides a sandbox to prevent some common unexpected bugs.

> lumen-roadrunner is a fork of [spiral/roadrunner-laravel](https://github.com/spiral/roadrunner-laravel)

#### Sandbox container

All the requests are processed in a sandbox of App Container. Any changes to `Application` class instance will not affect the initial Lumen application. Application bindings will also be resolved in every request. That means `singleton` instances only exist in the current request, and will not interfere with other requests.
However some instances can be marked to be pre-resolved in the beginning of the Worker's loop to prevent any unnecessary repeated resolving. This can be configured in the `roadrunner.php` configuration while or in the `AppServiceProvider` class of the app:
- services to warm: the list of instances to keep in memory
- services to clear: the list of instances to clear out after each request

For example instances handling database connections with pooling  might be worthwhile to keep in memory.

> Disclaimer: the above explanation about object lifecycles under RoadRunner is inspired by `laravel-swoole` 's [documentation](https://github.com/swooletw/laravel-swoole/wiki/2.-Swoole-Structure).

irfan-dahir commented:
@pushrbx Awesome work! 🤩 This is some serious effort. It's going to take me a bit to absorb these changes as my PHP has gotten a bit rusty. But thank you for leaving a detailed explanation and I've also seen you've left some comments within the code. So that really helps!

Due to the massive refactor, I'm breaking down my review into several parts.
- [ ] Code Review (Reviewing file changes via GitHub review). Almost there! I've gone through 400~ files 😅 ETA: 22/2/23
- [ ] Set up staging server (I'll be setting it up to test out the endpoints. Will share the access here.) ETA: 24/2/23
- [ ] Testing endpoints ETA: 24/2/23
- [ ] Reviewing tests and integration testing ETA: 25/2/23
- [ ] In-Depth Code Review (I'll be syncing with changes locally and understanding the changes in-depth. This won't be a blocker for this PR) ETA: N/A

I will not be covering Docker at the moment as I don't have experience there so I trust you with that. 😄

irfan-dahir reviewed (approved):
Code LGTM! There were a few fixme's in there btw. I presumed they're not for this PR so I've skipped over them.

I'll be setting up the staging server next.

pushrbx commented:
most fixme -s are related to:
- roadrunner shenanigans
- the date time accuracy in mongo, because currently in many places it's done based on string. (not sure jensenngers lib converts them or not) There were a few bugs because of this.

irfan-dahir commented on app/Anime.php:
```diff
@@ -259,12 +259,13 @@ public function typesenseQueryBy(): array
             'title_english_transformed',
             'title_japanese',
             'title_japanese_transformed',
+            'title_synonyms',
```
@pushrbx We removed synonyms because they were degrading results. Should we be adding it back?

pushrbx replied:
Yes. While testing with a wider range of people we found that this improves the hit results.
The most noteable edge case is the situation where the datasheet on mal doesn't have an english title in gray under the original title, so the parser just falls back to the original title so the english title will be same as the original one, however the actual english title appears as a synonym/alternative title. You can observe this with `Utawarerumono`, you can't search for its english title `The One Being Sung`.
Similar situation for `Slayers`: if the community decides to add localised titles for it e.g. the hungarian `Kis boszorkány` then people using the api from that country would want to be able to search for the localised title.
Worth noting with this change I might have broke something else. Or last time we didn't have the correct weights for ranking.

irfan-dahir replied:
Fair enough.

> Worth noting with this change I might have broke something else. Or last time we didn't have the correct weights for ranking.

Could be, we'll have to keep an eye on it. I don't remember what was diminishing the results last time with synonyms. Could've been incorrect weights too.

Staging currently has this change so I'll test it out further there.

irfan-dahir commented:
@Wamy-Dev That would be great. Please share any issues you come across here or our discord channel: https://discord.com/channels/460491088004907029/1085070803432919121

pushrbx commented on app/Services/TypeSenseScoutSearchService.php:
```diff
@@ -65,6 +65,11 @@ public function search(string $q, ?string $orderByField = null,
                 }
 
+                // if order_by is a user supplied value make it a priority over _text_match
+                if ($orderByField !== 'members' && !is_null($orderByField)) {
```
this `$oderByField !== 'members'` doesn't belong here, this should belong to another class/trait which exposes the method to apply things like this.

irfan-dahir replied:
@pushrbx could you show me how this would be done?

pushrbx replied:
This entirely belongs to the model I think. So the anime model's `getSearchIndexSortBy()` function's signature should probably change from
```php
public function getSearchIndexSortBy(): array|null
```
to
```php
public function getSearchIndexSortBy($orderByField, $sortDirectionDescending): array|null
```
and implement this logic accordingly. If the logic is same for both anime and manga, the logic should live in a service class i guess, and resolve the instance of that class inside the model. I can change this unless you want to take this on yourself. 😄

irfan-dahir replied:
@pushrbx the logic would be same for anime and manga, yep. So would need that service class.

If you could make this change, that would be very helpful 🙏

pushrbx replied:
Will do 👍

pushrbx replied:
I'm going to resolve this in a more simpler way.
By default when the user searches we want to show hits by relevancy, and then order them by "members".
When the `order_by` parameter is present, we don't want to show hits by relevancy anymore, the hits should be ordered by the field in the `order_by` parameter, and the relevancy score should be used for tie breaking.

So this bit:

```php
// override ordering field
if (!is_null($orderByField)) {
    $options['sort_by'] = "_text_match:desc,$orderByField:" . ($sortDirectionDescending ? "desc" : "asc");
}

// if order_by is a user supplied value make it a priority over _text_match
if ($orderByField !== 'members' && !is_null($orderByField)) {
    $options['sort_by'] = "$orderByField:" . ($sortDirectionDescending ? "desc" : "asc") . ",_text_match:desc";
}
```
Changes to this:
```php
// override ordering field
if (!is_null($orderByField)) {
    $options['sort_by'] = "$orderByField:" . ($sortDirectionDescending ? "desc" : "asc") . ",_text_match:desc";
}
```

pushrbx commented on app/Http/HttpHelper.php:
```diff
@@ -70,7 +70,7 @@ public static function serializeEmptyObjectsControllerLevel(array $data): array
     public static function getRouteName(Request $request) : string
     {
-        $route = explode('\\', $request->route()[1]['uses']);
+        $route = explode('\\', $request->route()[1]['uses'] ?? 'App\Undefined');
```
Afaik route is empty for a reason, and it should throw an exception and we should investigate why it happens. This behaves differently depending whether you run the app through artisan directly or as a PSR worker. I've fixed a bug recently which caused this bit to throw an exception.

irfan-dahir replied:
@pushrbx Should be fine to return Undefined in this instance though?

pushrbx replied:
I think I've missed this comment. The goal here is to get an exception at runtime with the entirely empty value, or at least it should throw an error if `$request->route()[1]['uses']` is null, because that's a fatal error, it should never be empty. If it's empty then we've made a programming error / programmer error somewhere, so we need a canary against it. 😄

pushrbx commented:
Also I've found that the elasticsearch bit of the code is completely borked: The `jikanPaginate` scout mixin is mostly specific to typesense and doesn't work with elasticsearch. So I think the elasticsearch support should be removed, and just focus on typesense for now. However we should keep it "robust" / "pluggable", so if anyone wants to contribute they could easily add an implementation.

pushrbx commented:
Fixed the stuff we highlighted after testing in staging:

- genre filtering
- added unapproved filtering
- fixed sfw filtering
- fixed kids filtering
- fixed deprecation warnings as of php 8.1
  - DateTime ctor can't take `null` anymore as first param
  - preg_replace doesn't accept `null` values as `$subject`
- fixed failed items indexer (the --failed option of indexers) -- it didn't load the correct file making it impossible to retry the indexing
- changed the document schema for search indexed anime/manga
  - added `approved` field to them
  - staging will require a reimport into TypeSense
- the central filtering system will now process `sfw` and `unapproved` filters, so they will be applied implicitly through the `filter` model scope method.

Accordingly the seasons, schedule, random, and search endpoints need testing again on staging.

irfan-dahir commented:
@pushrbx LGTM. Wanna do the honors?

Prid13 commented:
Amazing work 👏 So it's done? 😇 When will it be pushed to the live rest api?
